rs/ledger_suite/icp/tests/golden_nns_state.rs

```diff
@@ -19,74 +25,147 @@ const INDEX_CANISTER_ID: CanisterId =
 /// ledger canister suite.
 #[test]
 fn should_create_state_machine_with_golden_nns_state() {
-    let state_machine = new_state_machine_with_golden_nns_state_or_panic();
+    let setup = Setup::new();
+
+    // Verify ledger, archives, and index block parity
+    setup.verify_ledger_archive_index_block_parity();
 
     // Upgrade all the canisters to the latest version
-    upgrade_index(&state_machine, build_ledger_index_wasm().bytes());
-    upgrade_ledger(&state_machine, build_ledger_wasm().bytes());
-    upgrade_archive_canisters(&state_machine, build_ledger_archive_wasm().bytes());
+    setup.upgrade_to_master();
+    // Upgrade again to test the pre-upgrade
+    setup.upgrade_to_master();
+
+    // Verify ledger, archives, and index block parity
+    setup.verify_ledger_archive_index_block_parity();
 
     // Downgrade all the canisters to the mainnet version
-    upgrade_archive_canisters(&state_machine, build_mainnet_ledger_archive_wasm().bytes());
-    upgrade_ledger(&state_machine, build_mainnet_ledger_wasm().bytes());
-    upgrade_index(&state_machine, build_mainnet_ledger_index_wasm().bytes());
+    setup.downgrade_to_mainnet();
+
+    // Verify ledger, archives, and index block parity
+    setup.verify_ledger_archive_index_block_parity();
 }
 
-fn list_archives(state_machine: &StateMachine) -> Archives {
-    Decode!(
-        &state_machine
-            .query(LEDGER_CANISTER_ID, "archives", Encode!().unwrap())
-            .expect("failed to query archives")
-            .bytes(),
-        Archives
-    )
-    .expect("failed to decode archives response")
+struct Wasms {
+    ledger: Wasm,
+    index: Wasm,
+    archive: Wasm,
 }
 
-fn upgrade_archive(
-    state_machine: &StateMachine,
-    archive_canister_id: CanisterId,
-    wasm_bytes: Vec<u8>,
-) {
-    state_machine
-        .upgrade_canister(archive_canister_id, wasm_bytes, vec![])
-        .unwrap_or_else(|e| {
-            panic!(
-                "should successfully upgrade archive '{}' to new local version: {}",
-                archive_canister_id, e
-            )
-        });
+struct Setup {
+    state_machine: StateMachine,
+    master_wasms: Wasms,
+    mainnet_wasms: Wasms,
 }
 
-fn upgrade_archive_canisters(state_machine: &StateMachine, archive_wasm_bytes: Vec<u8>) {
-    let archives = list_archives(state_machine).archives;
-    for archive_info in &archives {
-        upgrade_archive(
+impl Setup {
+    pub fn new() -> Self {
+        let state_machine = new_state_machine_with_golden_nns_state_or_panic();
+
+        let master_wasms = Wasms {
+            ledger: build_ledger_wasm(),
```

Comment:
There are multiple wasm versions due to the memory upgrade. Which one do we want to use?

Reply:
This would use the "main" version, i.e., the tip of master without any additional features flags. For the memory upgrade, we do have an additional ledger wasm being built, but for V1 of the ledger that additional wasm has functionality that we would only release in V2, and for that testing we currently rely on the "regular" `StateMachine` tests in the V1 PR. Once V1 is merged to master and deployed to mainnet, we would test e.g., the V2 to V1 downgrade in this test also, and in case we uncover issues, we could still fix them in a V1.1 release.